### PR TITLE
Variable max header size

### PR DESCRIPTION
**Pull request by jakubkulhan**
`RequestHeaderParser` has hard-wired 4096 bytes for header buffer. However single cookie can easily eat up whole buffer.

### CODE REVIEW

**Comment by cboden** [comment]
LGTM

File: src/Server.php [hunk 1]
```diff
@@ -11,16 +11,19 @@ class Server extends EventEmitter implements ServerInterface
 {
     private $io;
 
-    public function __construct(SocketServerInterface $io)
+    private $maxHeaderSize;
+
+    public function __construct(SocketServerInterface $io, $maxHeaderSize = 4096 /* 4 KiB */)
```
Comment (clue) [inline src/Server.php]: The `maxHeaderSize` is not a property of the `Server` (this violates a SOLID design/DDD).

Ideally, we should inject a `RequestHeaderParser` via DI to the server (and make it default to the current behavior).

Unfortunately, the `RequestHeaderParser` is stateful, so we can't reuse it for multiple (concurrent) client connections. 

Perhaps we should consider passing a single instance to the ctor and using a clone thereof for each connection?

Arguably, this isn't necessarily part of this changeset. However, given that your changes introduce the necessity to pass this in the first place, this might make sense to discuss here.

Any thoughts?

File: src/Server.php [hunk 2]
```diff
@@ -41,6 +44,9 @@ public function __construct(SocketServerInterface $io)
                     $conn->emit('resume');
                 });
             });
+            $parser->on('error', function (\Exception $e) {
+                $this->emit('error', [$e]);
```
Comment (clue) [inline src/Server.php]: How is this related to your above change? At the very least, this needs some documentation :+1:

**Comment by clue** [comment]
Thanks @jakubkulhan, makes perfect sense to get this in! :+1:

File: src/Server.php (hunk 1 above)
Comment (jakubkulhan) [inline src/Server.php]: Then it might make sense to create `RequestHeaderParserFactory` that gets injected and creates `RequestHeaderParser`s for request (internally it can use cloning). However this might be a little bit of over-engineering.

**Comment by WyriHaximus** [comment]
Looks mostly good to me but I agree with @clue's comments :+1:

**Comment by HonzaMac** [comment]
@jakubkulhan Hi, are you going to finish this PR?

**Comment by jakubkulhan** [comment]
@HonzaMac Sorry, I don't have time for this at the moment. If you want to, you can take up the code I've written so far and complete it with suggested refactorings.
